modules/transactionpool/update.go

```diff
@@ -33,17 +36,87 @@ func (tp *TransactionPool) ProcessConsensusChange(cc modules.ConsensusChange) {
 				tp.log.Println("ERROR: could not delete a transaction:", err)
 			}
 		}
+
+		// Pull the transactions out of the fee summary. For estimating only
+		// over 10 blocks, it is extermely likely that there will be more
+		// applied blocks than reverted blocks, and if there aren't (a height
+		// decreasing reorg), there will be more than 10 applied blocks.
+		if len(tp.txnsPerBlock) > 0 {
+			tailOffset := uint64(0)
+			for i := 0; i < len(tp.txnsPerBlock)-1; i++ {
+				tailOffset += tp.txnsPerBlock[i]
+			}
+
+			// Strip out all of the transactions in this block.
+			tp.recentConfirmedFees = tp.recentConfirmedFees[:tailOffset]
+			// Strip off the tail offset.
+			tp.txnsPerBlock = tp.txnsPerBlock[:len(tp.txnsPerBlock)-1]
+		}
 	}
 	for _, block := range cc.AppliedBlocks {
 		if tp.blockHeight > 0 || block.ID() != types.GenesisID {
 			tp.blockHeight++
 		}
 		for _, txn := range block.Transactions {
-			err := tp.addTransaction(tp.dbTx, txn.ID())
+			err := tp.putTransaction(tp.dbTx, txn.ID())
 			if err != nil {
 				tp.log.Println("ERROR: could not add a transaction:", err)
 			}
 		}
+
+		// Add the transactions from this block.
+		var totalSize uint64
+		for _, txn := range block.Transactions {
+			var feeSum types.Currency
+			size := uint64(len(encoding.Marshal(txn)))
+			for _, fee := range txn.MinerFees {
+				feeSum = feeSum.Add(fee)
+			}
+			feeAvg := feeSum.Div64(size)
+			tp.recentConfirmedFees = append(tp.recentConfirmedFees, feeSummary{
+				Fee:  feeAvg,
+				Size: size,
+			})
+			totalSize += size
+		}
+		// Add an extra zero-fee tranasction for any unused block space.
+		remaining := types.BlockSizeLimit - totalSize
+		tp.recentConfirmedFees = append(tp.recentConfirmedFees, feeSummary{
+			Fee:  types.ZeroCurrency,
+			Size: remaining, // fine if remaining is zero.
+		})
+		// Mark the number of fee transactions in this block.
+		tp.txnsPerBlock = append(tp.txnsPerBlock, uint64(len(block.Transactions)+1))
```

Comment:
Same here, feels like a lot of code when all we're doing is adding all of transactions in the recent block into the set of recent transactions.